File: api/schemas/goals.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, Field, model_validator


GoalStatusEnum = Literal[
    "active", "achieved", "abandoned", "paused", "completed"
]
GoalStatusStored = Literal["active", "achieved", "abandoned", "paused"]


class GoalCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    target_amount: Decimal = Field(..., gt=0)
    target_currency: Literal["CRC", "USD"] = "CRC"
    target_date: Optional[date] = None
    # Legacy input alias from pre-6e API.
    deadline: Optional[date] = None
    current_amount: Decimal = Field(Decimal("0"), ge=0)
    priority: int = Field(3, ge=1, le=5)
    monthly_contribution: Optional[Decimal] = Field(None, gt=0)
    linked_account_id: Optional[uuid.UUID] = None

    @model_validator(mode="after")
    def fill_target_date_from_legacy_deadline(self) -> "GoalCreate":
        if self.target_date is None and self.deadline is not None:
            self.target_date = self.deadline
        return self


class GoalUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    target_amount: Optional[Decimal] = Field(None, gt=0)
    target_currency: Optional[Literal["CRC", "USD"]] = None
    current_amount: Optional[Decimal] = Field(None, ge=0)
    target_date: Optional[date] = None
    # Legacy input alias from pre-6e API.
    deadline: Optional[date] = None
    priority: Optional[int] = Field(None, ge=1, le=5)
    monthly_contribution: Optional[Decimal] = Field(None, gt=0)
    status: Optional[GoalStatusEnum] = None
    linked_account_id: Optional[uuid.UUID] = None


class GoalResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    name: str
    target_amount: Decimal
    target_currency: str
    current_amount: Decimal
    target_date: Optional[date]
    # Kept in responses so older clients do not break during the 6e rollout.
    deadline: Optional[date] = None
    monthly_contribution: Optional[Decimal]
    priority: int
    status: str
    linked_account_id: Optional[uuid.UUID]
    created_at: datetime

    model_config = {"from_attributes": True}


class GoalContributionCreate(BaseModel):
    amount: Decimal = Field(..., gt=0)
    occurred_at: Optional[datetime] = None
    transaction_id: Optional[uuid.UUID] = None


class GoalContributionResponse(BaseModel):
    id: uuid.UUID
    goal_id: uuid.UUID
    transaction_id: Optional[uuid.UUID]
    amount: Decimal
    occurred_at: datetime
    created_at: datetime

    model_config = {"from_attributes": True}


class GoalContributionResult(BaseModel):
    goal: GoalResponse
    contribution: GoalContributionResponse


class GoalProgress(BaseModel):
    id: uuid.UUID
    name: str
    target_amount: Decimal
    target_currency: str
    current_amount: Decimal
    remaining: Decimal
    progress_percent: float
    months_remaining: Optional[int]
    monthly_needed: Optional[Decimal]
    on_track: Optional[bool]
    status: str


class ContributeRequest(GoalContributionCreate):
    """Legacy request body for POST /goals/{id}/contribute."""


class GoalForecastResponse(BaseModel):
    """Phase 6e B9 — simple "at this pace, you arrive in X months" forecast.

    `has_enough_data` is False when there are zero contributions in the last
    3 full calendar months (we don't extrapolate from nothing). Already-
    achieved goals return `months_to_target=0` and `remaining=0`.
    """

    goal_id: uuid.UUID
    target_amount: Decimal
    current_amount: Decimal
    remaining: Decimal
    avg_monthly_contribution: Decimal
    months_to_target: Optional[int]
    projected_completion_date: Optional[date]
    has_enough_data: bool
    lookback_months: int = 3

File: web/src/api/goals.ts
import { api } from "./client";
import {
  GoalContribution,
  GoalContributionList,
  GoalDetail,
  GoalDetailList,
  GoalForecast,
} from "../schemas/entities";

export async function fetchGoals(status?: string) {
  const resp = await api.get("/goals", {
    params: status ? { status } : undefined,
  });
  return GoalDetailList.parse(resp.data);
}

export async function fetchGoal(id: string) {
  const resp = await api.get(`/goals/${id}`);
  return GoalDetail.parse(resp.data);
}

export async function fetchGoalContributions(id: string) {
  const resp = await api.get(`/goals/${id}/contributions`);
  return GoalContributionList.parse(resp.data);
}

export async function fetchGoalForecast(id: string) {
  const resp = await api.get(`/goals/${id}/forecast`);
  return GoalForecast.parse(resp.data);
}

export interface GoalCreatePayload {
  name: string;
  target_amount: string;
  target_currency: "CRC" | "USD";
  target_date?: string;
  linked_account_id?: string;
  priority?: number;
  monthly_contribution?: string;
}

export async function createGoal(payload: GoalCreatePayload) {
  const resp = await api.post("/goals", payload);
  return GoalDetail.parse(resp.data);
}

export interface GoalUpdatePayload {
  name?: string;
  target_amount?: string;
  target_date?: string | null;
  monthly_contribution?: string | null;
  priority?: number;
  status?: "active" | "achieved" | "abandoned" | "paused";
  linked_account_id?: string | null;
}

export async function updateGoal(id: string, payload: GoalUpdatePayload) {
  const resp = await api.patch(`/goals/${id}`, payload);
  return GoalDetail.parse(resp.data);
}

export async function abandonGoal(id: string) {
  return updateGoal(id, { status: "abandoned" });
}

export async function pauseGoal(id: string) {
  return updateGoal(id, { status: "paused" });
}

export async function resumeGoal(id: string) {
  return updateGoal(id, { status: "active" });
}

export async function markGoalAchieved(id: string) {
  return updateGoal(id, { status: "achieved" });
}

export interface GoalContributionPayload {
  amount: string;
  occurred_at?: string;
  transaction_id?: string;
}

export async function addGoalContribution(
  goalId: string,
  payload: GoalContributionPayload,
) {
  const resp = await api.post(`/goals/${goalId}/contributions`, payload);
  return resp.data as { goal: GoalDetail; contribution: GoalContribution };
}
